tests/tls_e2e/common/utility.cpp

```diff
@@ -0,0 +1,318 @@
+        uint8_t tmp = modulus[i];
+        modulus[i] = modulus[modulus_size - 1 - i];
+        modulus[modulus_size - 1 - i] = tmp;
+    }
+
+    // Calculate the MRSIGNER value which is the SHA256 hash of the
+    // little endian representation of the public key modulus. This value
+    // is populated by the signer_id sub-field of a parsed oe_report_t's
+    // identity field.
+
+    OE_TRACE_ERROR("modulus_size=%ld\n", modulus_size);
+    if (Sha256(modulus, modulus_size, signer) != 0)
+    {
+        OE_TRACE_ERROR("Sha256 failed\n");
+        goto done;
+    }
+
+    if (memcmp(signer, signer_id_buf, signer_id_buf_size) != 0)
+    {
+        OE_TRACE_ERROR("mrsigner is not equal!\n");
+        for (size_t i = 0; i < signer_id_buf_size; i++)
+        {
+            OE_TRACE_ERROR(
+                "0x%x - 0x%x\n", (uint8_t)signer[i], (uint8_t)signer_id_buf[i]);
+        }
+        goto done;
+    }
+    ret = true;
+done:
+    if (signer)
+        free(signer);
+
+    if (modulus != NULL)
+        free(modulus);
+
+    mbedtls_pk_free(&ctx);
+    return ret;
+}
+
+// If set, the verify callback is called for each certificate in the chain.
+// The verification callback is supposed to return 0 on success. Otherwise, the
+// verification failed.
+// The function should return 0 for anything (including invalid
+// certificates) other than fatal error, as a non-zero return
+// code immediately aborts the verification process.
+// For fatal errors, a specific error code should be used
+// (different from MBEDTLS_ERR_X509_CERT_VERIFY_FAILED which
+// should not be returned at this point), or MBEDTLS_ERR_X509_FATAL_ERROR
+// can be used if no better code is available.
+//
+int cert_verify_callback(
+    void* data,
+    mbedtls_x509_crt* crt,
+    int depth,
+    uint32_t* flags)
+{
+    oe_result_t result = OE_FAILURE;
+    int ret = 0;
+    unsigned char* cert_buf = NULL;
+    size_t cert_size = 0;
+
+    (void)data;
+
+    *flags = (uint32_t)MBEDTLS_ERR_X509_CERT_VERIFY_FAILED;
+    if (g_control_config.fail_cert_verify_callback)
+    {
+        OE_TRACE_INFO(
+            "Purposely returns failure from server's cert_verify_callback()\n");
+        goto done;
+    }
+
+    cert_buf = crt->raw.p;
+    cert_size = crt->raw.len;
+
+    OE_TRACE_INFO("cert_verify_callback with depth = %d\n", depth);
+    OE_TRACE_INFO("crt->version = %d\n", crt->version);
+    OE_TRACE_INFO("cert_size = %zu\n", cert_size);
+
+    if (cert_size <= 0)
+        goto done;
+
+    OE_TRACE_INFO("Calling oe_verify_attestation_certificate\n");
+    if (g_control_config.fail_oe_verify_attestation_certificate)
+        goto done;
+
+    result = oe_verify_attestation_certificate(
+        cert_buf, cert_size, enclave_identity_verifier, NULL);
+    OE_CHECK_MSG(
+        result,
+        "oe_verify_attestation_certificate failed with result = %s\n",
+        oe_result_str(result));
+
+    OE_TRACE_INFO("\nReturned from oe_verify_attestation_certificate\n");
+    ret = 0;
```

Comment:
Maybe ret shoud be initialized to a failure value instead of 0?

Reply:
Good catch. Appropriate error code was set to MBEDTLS_ERR_X509_CERT_VERIFY_FAILED.